Incremental decommits off-chain

v0d1ch opened the pull request:
This PR introduces changes related to off-chain decommit protocol logic. 
Now user has two options to request a decommit of some UTxO from the open Head:
  - Issue a http request to `/decommit` endpoint 
  - Issue a client input `Decommit`

Decommit happens in sequence and you can't ask for a decommit while another one is _in flight_.
 
It is the user's responsibility to craft a decommit transaction which should be applicable to the local ledger state on `L2`.

If the transaction applies then the `ReqDec` network effect is issued and the next `Snapshot` that is to be signed specifies 
which `UTxO` is to be decommitted from the Head. 

On next `AckSn` all nodes need to agree/sign the snapshot that excludes the decommit `UTxO` from the ledger state. 

Once everybody signs the snapshot decommit tx is issued and upon observing it on `L1`  it is removed from the local state and another decommit can be processed.


---

<!-- Consider each and tick it off one way or the other -->
* [ ] CHANGELOG updated or not needed
* [ ] Documentation updated or not needed
* [ ] Haddocks updated or not needed
* [ ] No new TODOs introduced or explained herafter

ch1bo committed:
Draft an end-to-end happy path scenario for decommit

v0d1ch committed:
Use http call in decommit test

Remove some fixmes but also add one more in HeadLogic to remind us
to change the response to tx instead of UTxO.

v0d1ch committed:
Tackle the api schema and ServerOutput encoding

Once we figure out exact type we want to return we will resolve FIXME in
the api schema

v0d1ch committed:
Implement decrement function

Add HeadParameters to DecrementTx
Stub out the decrementTx with appropriate params
Rename maybeEmitDecommitApproved to maybeEmitDecrementTx

v0d1ch committed:
Add snapshot UTxO to DecrementTx and mostly implement decrementTx

decrementTx is missing the head redeemer for decrement so this is moving
into on-chain territory a bit.

TODO: Add Decrement redeemer
TODO: How to construct the correct UTxO hash

Only emit DecrementTx if leader

v0d1ch committed:
Add snapshot and signatures to DecrementTx

Construct the Decrement redeemer using provided signatures and
utxoHash using the UTxO from confirmed Snapshot.

v0d1ch committed:
Fix the api.yaml schema and golden files

v0d1ch committed:
Use empty ByteString in case decommit utxo is not there

v0d1ch committed:
Check if decrementTx produces valid tx

This test file will also serve as a placeholder for
mutation tests down the line.

Add dummy checkDecrement validator that just returns True for now.

v0d1ch committed:
Fix plutus golden files

Also tackle some warnings in code.

v0d1ch committed:
Formatting changes

ffakenz commented on hydra-cardano-api/src/Cardano/Api/UTxO.hs:
```diff
@@ -82,6 +82,14 @@ render (k, TxOut _ (txOutValueToValue -> v) _ _) =
+
+-- | Infix version of 'difference'.
+(\\) :: UTxO' out -> UTxO' out -> UTxO' out
+a \\ b = difference a b
```
this is not used.

feels weird to have to define this function twice.

i would keep only the infix version and define `difference` in the where clause.

wdyt?

v0d1ch replied:
We thought infix version might come in handy. We can certainly remove it if you want.

ffakenz commented on hydra-cluster/src/Hydra/Cluster/Scenarios.hs:
```diff
@@ -638,29 +638,33 @@ canDecommit tracer workDir node hydraScriptsTxId =
+            walletSk
+
+      send n1 $ input "Decommit" ["decommitTx" .= decommitTx]
       -- NOTE: Alternative:
```
do we want to keep this NOTE?

v0d1ch replied:
We should test both ways of decomitting, using client input and http endpoint to make sure they actually work. Merging this PR doesn't mean the work is over, there is still a lot of TODOs and FIXMEs so one NOTE is not really important imo.

v0d1ch replied:
I used `oneof` in the end to alter between the two ways of decommitting.

ffakenz commented on hydra-node/src/Hydra/API/ServerOutput.hs:
```diff
@@ -100,6 +100,10 @@ data ServerOutput tx
       , participants :: [OnChainId]
       }
   | DecommitRequested {headId :: HeadId, utxoToDecommit :: UTxOType tx}
+  | DecommitAlreadyInFlight {headId :: HeadId, decommitTx :: tx}
```
do not we need haddocks for this and the above outputs?

v0d1ch replied:
I added the description for these ctors in the `api.yaml`

ffakenz commented on hydra-node/src/Hydra/Chain/Direct/Handlers.hs:
```diff
@@ -324,6 +326,9 @@ convertObservation = \case
   CollectCom CollectComObservation{headId} ->
     pure OnCollectComTx{headId}
+  Decrement DecrementObservation{headId} ->
+    -- FIXME: implement decrement observation
```
are we planing to solve this FIXME in a future PR?

v0d1ch replied:
Yes of course. This PR is just a step forward into decommit direction and we are doing it in order for the team to have better visibilitty into the work done and we don't end up in the end with one big _do it all_ PR.

v0d1ch replied:
Actually, now this PR also contains the observation code so this fixme is resolved.

ffakenz commented on hydra-plutus/src/Hydra/Contract/Head.hs:
```diff
@@ -226,6 +227,10 @@ commitDatum input = do
 {-# INLINEABLE commitDatum #-}
 
+checkDecrement :: Bool
+checkDecrement = True
```
are we implementing this in a future PR?

v0d1ch replied:
Yes.

ffakenz commented on hydra-node/src/Hydra/Chain/Direct/Tx.hs:
```diff
@@ -314,6 +314,73 @@ collectComTx networkId scriptRegistry vk headId headParameters (headInput, initi
+      & addOutputs [headOutput', decrementOutput]
+      & addExtraRequiredSigners [verificationKeyHash vk]
+ where
+  -- TODO: add Decrement redeemer
```
I think we can remove this TODO already as it seems to have been addressed.

ffakenz commented on hydra-node/src/Hydra/Chain/Direct/Tx.hs:
```diff
@@ -314,6 +314,73 @@ collectComTx networkId scriptRegistry vk headId headParameters (headInput, initi
   commitRedeemer =
     toScriptData $ Commit.redeemer Commit.ViaCollectCom
 
+decrementTx ::
```
can we add a haddock to this function?

ffakenz commented on hydra-node/src/Hydra/HeadLogic/Error.hs:
```diff
@@ -38,12 +38,15 @@ data RequirementFailure tx
   | SnapshotDoesNotApply {requestedSn :: SnapshotNumber, txid :: TxIdType tx, error :: ValidationError}
+  | DecommitTxInvalid {decommitTx :: tx, error :: ValidationError}
+  | DecommitTxInFlight {decommitTx :: tx}
+  | DecommitDoesNotApply {decommitTx :: tx, error :: ValidationError}
```
what is the difference btw DecommitTxInvalid and DecommitDoesNotApply?

can't we just have a single and use it in both places?

v0d1ch replied:
Code changed after your comment so this isn't relevant anymore. Now we have one type which we return in case of any potential error.

ffakenz commented on hydra-node/src/Hydra/Snapshot.hs:
```diff
@@ -55,32 +59,50 @@ instance IsTx tx => FromJSON (Snapshot tx) where
       serialise (toData $ toBuiltin $ serialiseToRawBytes headId)
         <> serialise (toData $ toBuiltin $ toInteger number) -- CBOR(I(integer))
         <> serialise (toData $ toBuiltin $ hashUTxO @tx utxo) -- CBOR(B(bytestring)
+        -- TODO: is serializing separately fine or we should get the diff
```
do not get this question, is it still open?

v0d1ch replied:
It is about serialization of the decommit utxo and wether we should have it separate or serialize snapshot utxo with already removed decommit utxo.

ffakenz commented on hydra-node/test/Hydra/Chain/Direct/ContractSpec.hs:
```diff
@@ -112,6 +114,9 @@ spec = parallel $ do
       propMutation healthyCollectComTx genCollectComMutation
+  describe "Decrement" $ do
+    prop "is healthy" $
+      propTransactionEvaluates healthyDecrementTx
```
are we adding mutation specs for this in a future PR?

v0d1ch replied:
Yes. This should be right after we have the on-chain code since mutations test the on-chain part.

ffakenz commented on hydra-node/test/Hydra/Chain/Direct/Contract/Contest.hs:
```diff
@@ -106,6 +106,7 @@ healthyContestSnapshot =
     , number = healthyContestSnapshotNumber
     , utxo = healthyContestUTxO
     , confirmed = []
+    , utxoToDecommit = Nothing
```
why not arbitrary?

v0d1ch replied:
We are not trying to test decommits here but in any case `utxoToDecommit` must be a subset of `utxo` from the snapshot so shouldn't have arbitrary value since ideally in tests we would decommit only what was committed before just like in real world (no on-chain code to enforce that yet)

ffakenz commented on hydra-node/test/Hydra/Chain/Direct/Contract/Close.hs:
```diff
@@ -148,6 +148,7 @@ healthySnapshot =
     , number = healthyCloseSnapshotNumber
     , utxo = healthyCloseUTxO
     , confirmed = []
+    , utxoToDecommit = Nothing
```
why not arbitrary?

v0d1ch replied:
Adding this field just to remove the compiler warning. We will test thoroughly the decommit workflow with respect to `Snapshot` type.

ffakenz commented on hydra-node/test/Hydra/BehaviorSpec.hs:
```diff
@@ -377,6 +377,42 @@ spec = parallel $ do
 
                 waitUntil [n1] $ GetUTxOResponse testHeadId (utxoRefs [2, 42])
 
+      it "can request decommit" $
```
i think this test is redundant as it is being covered by the below.

wdyt?

v0d1ch replied:
I think so too so we could remove it.

ffakenz commented on hydra-node/src/Hydra/Snapshot.hs:
```diff
@@ -157,7 +180,8 @@ genConfirmedSnapshot headId minSn utxo sks
     -- snapshots
     number <- arbitrary `suchThat` (> minSn)
-    let snapshot = Snapshot{headId, number, utxo, confirmed = []}
+    -- TODO: check whether we are fine with this not producing any decommitting utxo ever
```
are we suppose to have a spec for this check?

v0d1ch replied:
This todo is here to make sure we don't forget to check where `genConfirmedSnapshot` is used and if we want to actually test decommits using this function or it is ok to ignore them and leave them always as empty.

ffakenz commented on hydra-node/test/Hydra/HeadLogicSnapshotSpec.hs:
```diff
@@ -199,13 +200,14 @@ prop_singleMemberHeadAlwaysSnapshotOnReqTx sn = monadicST $ do
   pure $
     ( collectEffects outcome
-        `shouldContain` [NetworkEffect (ReqSn nextSn [txId tx])]
+        `shouldContain` [NetworkEffect (ReqSn nextSn [txId tx] Nothing)]
```
i see all tests in this module having Nothing as value for `decommitTx`.

should not we have a test where it is set to Just?

v0d1ch replied:
We just populated the new field to please the compiler. We will add tests for `ReqSn` with some decommit too.

v0d1ch replied:
You can take a look at the `HeadLogicSpec`, there we added some test using `ReqSn` with populated `decommitTx`.

ffakenz commented on hydra-node/test/Hydra/NodeSpec.hs:
```diff
@@ -146,6 +146,23 @@ spec = parallel $ do
+        node <- createHydraNode bobSk [alice, carol] defaultContestationPeriod events
+        (node', getNetworkMessages) <- recordNetwork node
+        runToCompletion tracer node'
+        getNetworkMessages `shouldReturn` [ReqSn{snapshotNumber = 1, transactionIds = [1], decommitTx = Nothing}]
```
i do not understand, why if it valid the ReqDec we have a Snapshot with Nothing in decommitTx?

v0d1ch replied:
Good point. This test was actually red since the logic behind it is already there and the decommitTx is populated. One thing I left a TODO for is why `transactionIds` is empty list instead of containing the tx id.

v0d1ch replied:
And in any case `transactionIds` should stay empty since we don't record what we put on L1.

v0d1ch committed:
Fix changes after rebase

v0d1ch committed:
Assert decommitTx is there in "Emits ReqSn on valid RecDec"

Also format the code

v0d1ch commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -581,6 +682,7 @@ onOpenClientClose st =
 --
 -- __Transition__: 'OpenState' → 'ClosedState'
 onOpenChainCloseTx ::
+  Monoid (UTxOType tx) =>
```
It is required by the `getSnapshot` and adding this constraint further up doesn't help us. Perhaps we can refactor this later since any general `UTxO` should have a `Monoid` instance.

v0d1ch committed:
Dump decommit observation code

v0d1ch committed:
Use correct address for decommit

Acidentally I was creating a decommit output at hydra-node
internal wallet address.

v0d1ch committed:
Draft a test to show we can only process decommits one by one

v0d1ch committed:
Make sure we process decommits in sequence

v0d1ch committed:
Formatting

v0d1ch committed:
Fix ChainObserverSpec with regards to decommits

v0d1ch committed:
Fix e2e spec adding a missing utxoToDecommit field

v0d1ch committed:
Move one test from e2e to HeadLogicSpec

v0d1ch committed:
Test both way of requesting a decommit

v0d1ch committed:
Some touches/refactoring related to pr comments

v0d1ch committed:
Yaml schema changes

v0d1ch committed:
Fix the api schema

ch1bo commented on hydra-cluster/src/Hydra/Cluster/Scenarios.hs:
```diff
@@ -661,16 +661,26 @@ canDecommit tracer workDir node hydraScriptsTxId =
+      waitFor hydraTracer 10 [n1] $
+        output "DecommitProcessed" ["headId" .= headId]
 
       failAfter 10 $ waitForUTxO node decommitUTxO
```
Should: expect `DecommitProcessed` after seing the UTxO become available (this line)

ch1bo commented on hydra-cluster/test/Test/EndToEndSpec.hs:
```diff
@@ -801,6 +816,7 @@ initAndClose tmpDir tracer clusterIx hydraScriptsTxId node@RunningNode{nodeSocke
             , "snapshotNumber" .= int expectedSnapshotNumber
             , "utxo" .= newUTxO
             , "confirmedTransactions" .= [txId tx]
+            , "utxoToDecommit" .= Null
```
Should: not require `null` values in our JSON encoding.

Either we would produce this as an empty object or not include the key at all. Using `null` is a bit of a middle ground (worst of both worlds).

BTW This test could be made robust against this, by not checking full equality on the snapshot, but only the fields we are interested in (e.g. `confirmedTransactions`)

ch1bo commented on hydra-node/golden/ReasonablySized (ServerOutput (Tx BabbageEra)).json:
Surprised that we have this *and* `golden/ReasonablySized (TimedServerOutput (Tx BabbageEra)).json`.

We could only keep one and be fine (and have smaller diffs)

v0d1ch replied:
The thing is, schema verification is using this file and everywhere in the api.yaml we have specified `seq`, `timestamp` in all of the types that have these fields so I refrained from changing it now.

ch1bo commented on hydra-node/json-schemas/api.yaml:
```diff
@@ -1734,6 +1888,10 @@ components:
             $ref: "api.yaml#/components/schemas/TxId"
+        utxoToDecommit:
+          oneOf:
+            - type: "null"
```
Must: make this consistent with `utxo` field in this same server output. i.e. use `{}` with empty UTxO

ch1bo commented on hydra-node/json-schemas/api.yaml:
```diff
@@ -147,7 +151,7 @@ channels:
       - localhost-http
     publish:
       description: Draft a decommit transaction, which needs to be signed by the requester and later on validated against the local ledger.
```
Should: update description

ch1bo commented on hydra-node/json-schemas/api.yaml:
```diff
@@ -156,11 +160,18 @@ channels:
           * 200 OK with empty body
           * 400 BadRequest on parsing errors with string as body
+      message:
+        $ref: "api.yaml#/components/messages/DecommitResponse"
```
Should: inline these message payloads (as they are not re-usable)

ch1bo replied:
Could: make the response behave consistent with other endpoints.

ch1bo commented on hydra-node/json-schemas/api.yaml:
Should: add some API tests in HTTPServerSpec

ch1bo commented on hydra-node/golden/Event SimpleTx.json:
A bit ouf of scope: Why do we have golden tests for `Event` type? We are not storing it anywhere (and certainly not as JSON). An interesting golden / regression test however would be to have the `StateChanged` type! (It would indicate breakage on the persisted `state` file)

ch1bo commented on hydra-node/hydra-node.cabal:
```diff
@@ -281,6 +281,7 @@ test-suite tests
     Hydra.Chain.Direct.Contract.CollectCom
     Hydra.Chain.Direct.Contract.Commit
     Hydra.Chain.Direct.Contract.Contest
+    Hydra.Chain.Direct.Contract.Decrement
```
This file does not exist (yet)

ch1bo commented on hydra-node/test/Hydra/BehaviorSpec.hs:
```diff
@@ -406,13 +406,38 @@ spec = parallel $ do
+                let decommitTx1 = SimpleTx 1 (utxoRef 1) (utxoRef 42)
+                send n2 (Decommit{decommitTx = decommitTx1})
+                waitUntil [n2] $
+                  DecommitRequestReceived{headId = testHeadId, utxoToDecommit = utxoRefs [42]}
```
Should: remove this as it does not actually provide any additional value and is one message less to explain

ch1bo commented on hydra-node/test/Hydra/BehaviorSpec.hs:
```diff
@@ -406,13 +406,38 @@ spec = parallel $ do
+
+                waitUntil [n1, n2] $ DecommitProcessed{headId = testHeadId}
+
+                send n1 (Decommit{decommitTx = decommitTx2})
```
Could: re-use the constructed `Decommit` client input value even.

ch1bo commented on hydra-node/json-schemas/api.yaml:
```diff
@@ -469,15 +481,64 @@ components:
+    DecommitTxInvalid:
+      title: DecommitTxInvalid
+      description: |
+        Decommit tx could not be applied to local ledger state.
```
Must: include the ledger error in this output. Otherwise a client application would not know why this failed.

We often also include the UTxO against which the validation failed (see [TxInvalid](https://github.com/input-output-hk/hydra/blob/9c7d20fe0720792b2fd8e1af2bfa470b6dd1620c/hydra-node/src/Hydra/API/ServerOutput.hs#L80))

ch1bo commented on hydra-node/src/Hydra/API/ServerOutput.hs:
```diff
@@ -99,11 +99,12 @@ data ServerOutput tx
+  | DecommitRequestReceived {headId :: HeadId, utxoToDecommit :: UTxOType tx}
   | DecommitRequested {headId :: HeadId, utxoToDecommit :: UTxOType tx}
+  | DecommitTxInvalid {headId :: HeadId, decommitTx :: tx}
   | DecommitAlreadyInFlight {headId :: HeadId, decommitTx :: tx}
```
Could: group the erroneous outputs related to requesting a decommit  

```suggestion
  | DecommitInvalid {headId :: HeadId, decommitInvalidReason :: DecommitInvalidReason}
  
  
data DecommitInvalidReason
  = DecommitTxInvalid { confirmedUTxO :: UTxOType tx, decommitTx :: tx, validationError :: ValidationError }
  | DecommitAlreadyInFlight
```

ch1bo commented on hydra-node/test/Hydra/HeadLogicSpec.hs:
```diff
@@ -209,6 +209,36 @@ spec =
+            _ -> False
+
+          let reqDecEvent = NetworkEvent defaultTTL bob ReqDec{transaction = decommitTx'}
+          -- REVIEW: do we need to keep track of decommit txs in transactionIds?
```
No

ch1bo commented on hydra-node/test/Hydra/HeadLogicSpec.hs:
```diff
@@ -209,6 +209,36 @@ spec =
             (Open OpenState{coordinatedHeadState = CoordinatedHeadState{decommitTx}}) -> decommitTx == Just decommitTx'
             _ -> False
 
+        it "Emits ReqSn on valid RecDec" $ do
```
```suggestion
        it "emits ReqSn on valid RecDec" $ do
```

ch1bo commented on hydra-node/test/Hydra/HeadLogicSpec.hs:
```diff
@@ -209,6 +209,36 @@ spec =
+          let reqSn = ReqSn{snapshotNumber = 1, transactionIds = [], decommitTx = Just decommitTx'}
+
+          let s1 = update bobEnv ledger s0 reqDecEvent
+          s1 `hasEffect` NetworkEffect reqSn
```
Should: implement this test or drop it

The body is a copy from above and the not tests what is intended above.

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -567,42 +567,40 @@ onOpenNetworkAckSn Environment{party} openState otherParty snapshotSignature sn
-
+  maybeEmitDecrementTx snapshot@Snapshot{utxoToDecommit} signatures outcome
+    | isJust decommitTx
+    , (utxoFromTx <$> decommitTx) == utxoToDecommit =
```
Checking this does not make sense here. If we would detect an inconsistency between the utxo to decommit in the snapshot (from our state) and the requested decommit tx (from our state), then it's already too late.

We **might want** to check this, but earlier. Let's not do it now and keep a big TODO in the ReqSn processiong `requireApplicableDecommitTx` function to maybe do this check there (and not sign the snapshot if something is off). We will revisit this with the spec again.

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -616,17 +614,20 @@ onOpenNetworkAckSn Environment{party} openState otherParty snapshotSignature sn
+--   - Alter it's state to record what is to be decommitted
+--   - Issue a server output 'DecommitRequested' with the relevant utxo
+--   - Issue a 'ReqSn' since all parties need to agree in order for decommit to
+--   be taken out of a Head.
```
Must: not send `ReqSn` if we are not the leader

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -815,6 +816,15 @@ update env ledger st ev = case (st, ev) of
     -- XXX: ttl == 0 not handled for AckSn
     onOpenNetworkAckSn env openState otherParty snapshotSignature sn
+  ( Open OpenState{headId = ourHeadId}
+    , OnChainEvent Observation{observedTx = OnDecrementTx{headId}}
```
No need to change anything, but maybe let's keep a note for later: This is exactly the location where we should consider _what happens if this observed tx is rolled back_ here.

ch1bo replied:
Suggest to re-order these to follow the natural flow of things: Client input > ReqDec > OnDecrementTx

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -828,19 +838,24 @@ update env ledger st ev = case (st, ev) of
-        [ ClientEffect ServerOutput.DecommitRequested{headId, utxoToDecommit}
-        , NetworkEffect ReqDec{transaction = decommitTx}
-        ]
+  (Open openState@OpenState{headId, coordinatedHeadState, currentSlot}, ClientEvent Decommit{decommitTx}) -> do
```
Could: move it into a top-level function like the other main steps of the protocol

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -828,19 +838,24 @@ update env ledger st ev = case (st, ev) of
-        , NetworkEffect ReqDec{transaction = decommitTx}
-        ]
+  (Open openState@OpenState{headId, coordinatedHeadState, currentSlot}, ClientEvent Decommit{decommitTx}) -> do
+    requireNoDecommitInFlight openState decommitTx $
```
Must: not send a `DecommitAlreadyInFlight` to clients of a `hydra-node` which did not issue the original `ReqDec` leading to this.

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -828,19 +838,24 @@ update env ledger st ev = case (st, ev) of
+            <> Error
+              ( RequireFailed $
+                  DecommitTxInvalid{decommitTx, error = err}
+              )
```
Should: not need to have the `Error RequireFailed` outcome and the `DecommitTxInvalid` server output. The server output alone is enough (it gets logged and we don't do anything neither on `Effects` and `Error` outcome anyway)

ch1bo commented on hydra-node/src/Hydra/Snapshot.hs:
```diff
@@ -78,10 +78,9 @@ instance forall tx. IsTx tx => SignableRepresentation (Snapshot tx) where
-        <> maybe mempty (serialise . (toData . toBuiltin . hashUTxO @tx)) utxoToDecommit -- CBOR(B(bytestring)
+        -- TODO: In the spec we have added new field but here we sign the diff
+        -- between two utxos.
+        <> serialise (toData $ toBuiltin $ hashUTxO @tx (utxo `withoutUTxO` maybe mempty id utxoToDecommit)) -- CBOR(B(bytestring)
```
Must: keep the serialization as it was before, i.e. hashing the `utxoToDecommit`

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/Decrement.hs:
```diff
@@ -0,0 +1,137 @@
+      parameters
+      (headInput, headOutput)
+      decommitUTxO
+      healthySnapshot
```
Should: avoid duplicated passing of utxo to decommit

ch1bo commented on hydra-node/src/Hydra/Chain.hs:
```diff
@@ -66,7 +67,13 @@ data PostChainTx tx
+      { headId :: HeadId
+      , headParameters :: HeadParameters
+      , decrementUTxO :: UTxOType tx
+      , snapshot :: Snapshot tx
```
Should: avoid duplication

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Tx.hs:
```diff
@@ -314,6 +314,65 @@ collectComTx networkId scriptRegistry vk headId headParameters (headInput, initi
+      (mkScriptAddress @PlutusScriptV2 networkId headScript)
+      (txOutValue headOutput)
+      headDatumAfter
+      ReferenceScriptNone
```
Should: avoid requiring a `networkId`, by just using `headOutput` and `modifyTxOutDatum` on it (like in the close)

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/Contract/Decrement.hs:
```diff
@@ -0,0 +1,137 @@
+{-# LANGUAGE DuplicateRecordFields #-}
+{-# OPTIONS_GHC -Wno-orphans #-}
```
No orphans here!

ch1bo reviewed (changes requested):
Fix the "must" ones before merging

v0d1ch committed:
Change Snapshot json encoding for utxoToDecommit

v0d1ch committed:
Changes to schema related to Snapshot field change

v0d1ch committed:
Changes to the api schema endpoints

We don't need to use dedicated $ref's

v0d1ch committed:
Add decommit http endpoint tests in HttpServerSpec

v0d1ch committed:
Add utxo and validationError to DecommitTxInvalid

We want better observability when something goes wrong

v0d1ch committed:
Introduce DecommitInvalid to hold invalid options

It makes sense to have one type for all possible invalid
outcomes.

v0d1ch committed:
Reuse utxoToDecommit from the Snapshot

We don't need to keep it in DecrementTx since we already have it
as part of a Snapshot

v0d1ch committed:
Do not require networkId in decrementTx

v0d1ch committed:
Remove decommit utxo check and add a TODO about it

v0d1ch committed:
Check if leader before outputting ReqDec

v0d1ch committed:
Revert signable rep for Snapshot

v0d1ch committed:
Remove DecomitRequestReceived

v0d1ch committed:
Remove orphans ignoring and some minor test changes

v0d1ch committed:
Output DecommitRecorded and DecommitRequested on each ReqDec

We only need to emit ReqSn if we are the leader, other events
should be published

v0d1ch committed:
Formatting changes

v0d1ch committed:
Reorder HeadLogic events to natural flow of decommit

v0d1ch committed:
Only emit DecommitAlreadyInFlight if we requested it

When we receive ReqDec it doesn't make sense to see
DecommitAlreadyInFlight if we didn't request the decommit
in the first place.

v0d1ch committed:
Fix the snapshot verification

Since we hash the utxo to decommit separately we also need another
argument to the verification function.

Also rename Decrement redeemer field to match the others

v0d1ch committed:
Simplify e2e test a bit

v0d1ch committed:
Fix plutus golden tests

ch1bo commented on hydra-node/src/Hydra/API/ServerOutput.hs:
```diff
@@ -49,6 +49,26 @@ instance IsChainState tx => FromJSON (TimedServerOutput tx) where
+  | DecommitAlreadyInFlight {decommitTx :: tx}
+  deriving stock (Generic)
+
+deriving stock instance IsChainState tx => Eq (DecommitInvalidReason tx)
```
Could use fine grained constraints to make this less demanding:

```suggestion
deriving stock instance (Eq tx, Eq (UTxoType tx)) => Eq (DecommitInvalidReason tx)
```

And the following instances below accordingly..

ch1bo commented on hydra-node/src/Hydra/API/ServerOutput.hs:
```diff
@@ -100,10 +120,9 @@ data ServerOutput tx
-    DecommitApproved {headId :: HeadId, utxoToDecommit :: UTxOType tx}
+  | DecommitInvalid {headId :: HeadId, decommitInvalidReason :: DecommitInvalidReason tx}
+  | DecommitApproved {headId :: HeadId, utxoToDecommit :: UTxOType tx}
+  | DecommitProcessed {headId :: HeadId}
```
Should: rename this to be less ambiguous

"Processed" does not really say that it's _done_ or _completed_. Relevant for the Hydra clients is, that after seeing this, they can expect that `POST /decommit` is possible again.

ch1bo replied:
Also, if we rename this, it would not conflict with the `StateChanged` data constructor

ch1bo commented on hydra-node/json-schemas/api.yaml:
```diff
@@ -146,21 +148,44 @@ channels:
+              properties:
+                tag:
+                  type: string
+                  enum: ["TransactionSubmitted"]
```
Could: Re-use the schema if we re-use the data type.

However, I'm not sure if this is a good response. It suggests that a transaction was submitted (onto L1; that's at least what we do when we used this message in `/cardano-transaction`). However, a positive response here is fully asynchronous to the submission of the decommit request. It could even be that we see a `DecommitInvalid` (already in flight) just after responding this.

Also, I'm not a fan of this singleton response type. It's not providing any additional information and having a description here would be more than enough (with a free form schema on the body / empty body)

ch1bo commented on hydra-node/json-schemas/api.yaml:
```diff
@@ -146,21 +148,44 @@ channels:
+                  enum: ["TransactionSubmitted"]
+          - title: DecommitError
+            payload:
+              type: string
```
Could: explain when this is returned (as you seem to want to describe it)

This is returned if the request body could not be parsed? Mentioning this would be valuable. But also fine if you don't mention it at all (e.g. on `/commit` we did not bother to describe the error responses at all)

ch1bo commented on hydra-node/json-schemas/api.yaml:
```diff
@@ -494,35 +547,92 @@ components:
+      title: DecommitTxInvalid
+      description: |
+        Requested decommit was rejected since it could not be applied to local ledger state.
+      payload:
```
Should: not put an arbitrary `payload` here if you don't use it!

Also, as this schema is only referenced from one location, I would suggest to inline it there. Doing that will highlight that this `payload` field is actually not needed.

The same holds for `DecommitAlreadyInFlight` schema below.

v0d1ch replied:
I swear I tried dong that and the validator was spitting out some weird errors....Probably because I started the payload with `title` or `name`. If I just put `type: object` and enumerate fields it works.

ch1bo commented on hydra-node/json-schemas/api.yaml:
```diff
@@ -494,35 +547,92 @@ components:
+    DecommitAlreadyInFlight:
+      title: DecommitAlreadyInFlight
+      description: |
+        Requested decommit was rejected since the previous decommit request is not yet finalized.
```
Note that this is a hint that naming the server output to wait for `DecommitFinalized` would be a good idea :)

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -1150,23 +1161,63 @@ recoverState = foldl' aggregate
   OpenState tx ->
-  tx ->
+  -- | We are the decommit request party
+  Bool ->
```
Must: not do this 

Boolean blindness is a thing we want to avoid, see our [Coding Standards](https://github.com/input-output-hk/hydra/wiki/Coding-Standards#avoid-boolean-blindness)

I would really avoid this function at all. There is even no re-use if we implement it correctly:

* on client input `DecommitTx` and `decommitTx` is not `Nothing` -> send server output `DecommitInvalid` with `DecommitAlreadyInFlight`
* on network messsage `ReqDec` and `decommitTx` is not `Nothing` -> log a `RequireFailed` and **no server output**

So please move this check into the respective places and not pass `Bool` around.

v0d1ch replied:
If it is in the coding standards bible who I am to object :)

ch1bo reviewed (changes requested):
I was _VERY_ close to approve this. Only the head logic was slightly off. It's a must change as I think we introduce these changes with this PR and make it behave similar to `NewTx` / `TxInvalid` behavior (+ avoid boolean blindness tech debt right away).

v0d1ch committed:
Inline requireNoDecommitInFlight

On ReqDec if there is a decommit tx present in state
just issue RequirementFailure DecommitTxInFlight

v0d1ch committed:
Rename DecommitProcessed to DecommitFinalized

v0d1ch committed:
Inline DecommitAlreadyInFlight and DecommitTxInvalid schemas

v0d1ch committed:
Don't reuse TransactionSubmitted for decommit

v0d1ch committed:
Use less demanding constraints for DecommitInvalidReason

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -826,26 +834,41 @@ update env ledger st ev = case (st, ev) of
+    case mExistingDecommitTx of
+      Just existingDecommitTx ->
+        Error $ RequireFailed $ DecommitTxInFlight{decommitTx = existingDecommitTx}
+      Nothing -> onOpenNetworkReqDec env openState transaction
```
Could: keep the whole logic in the handler function or none (to stay consistent with the other handlers)

I was caught off guard when reviewing this that this check is here and not in `onOpenNetworkReqDec`

ch1bo commented on hydra-node/src/Hydra/Snapshot.hs:
```diff
@@ -79,9 +83,7 @@ instance forall tx. IsTx tx => SignableRepresentation (Snapshot tx) where
-        -- TODO: is serializing separately fine or we should get the diff
-        -- between two utxos and sign that? In the spec we have added new field
-        <> maybe mempty (serialise . (toData . toBuiltin . hashUTxO @tx)) utxoToDecommit -- CBOR(B(bytestring)
+        <> serialise (maybe (toData $ toBuiltin $ hashUTxO @tx mempty) (toData . toBuiltin . hashUTxO @tx) utxoToDecommit) -- CBOR(B(bytestring)
```
Should: simplify this

```suggestion
        <> serialise (toData $ toBuiltin $ hashUTxO @tx $ fromMaybe mempty utxoToDecommit) -- CBOR(B(bytestring)
```

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Tx.hs:
```diff
@@ -314,6 +314,60 @@ collectComTx networkId scriptRegistry vk headId headParameters (headInput, initi
+  -- | Decrement UTxO.
+  UTxO ->
+  -- | Confirmed Snapshot
+  Snapshot Tx ->
```
Should: remove duplication between `UTxO` and `Snapshot`

the `utxoToDecommit` are already included in the `Snapshot`

v0d1ch replied:
For the record I wanted to avoid dealing with `Maybe` in the `decrementTx` but it should be easy as using `maybeToList`

v0d1ch committed:
Final changes
